src/batch/src/executor/row_seq_scan.rs

```diff
@@ -214,26 +211,59 @@ impl BoxedExecutorBuilder for RowSeqScanExecutorBuilder {
                 distribution,
             );
             let keyspace = Keyspace::table_root(state_store.clone(), &table_id);
-            let scan_type = if pk_prefix_value.size() == 0 && is_full_range(&next_col_bounds) {
+
+            if seq_scan_node.scan_ranges.is_empty() {
                 let iter = table.batch_dedup_pk_iter(source.epoch, &pk_descs).await?;
-                ScanType::TableScan(iter)
-            } else if pk_prefix_value.size() == pk_len {
-                let row = {
-                    keyspace.state_store().wait_epoch(source.epoch).await?;
-                    table.get_row(&pk_prefix_value, source.epoch).await?
+                return Ok(Box::new(RowSeqScanExecutor::new(
+                    table.schema().clone(),
+                    vec![ScanType::TableScan(iter)],
+                    RowSeqScanExecutorBuilder::DEFAULT_CHUNK_SIZE,
+                    source.plan_node().get_identity().clone(),
+                    batch_stats,
+                )));
+            }
+
+            let mut futures = vec![];
+            for scan_range in &seq_scan_node.scan_ranges {
+                let pk_types = pk_types.clone();
+                let table = table.clone();
+                let keyspace = keyspace.clone();
+
+                let get_scan_type = async move || {
+                    let (pk_prefix_value, next_col_bounds) =
+                        get_scan_bound(scan_range.clone(), pk_types.into_iter());
+
+                    let scan_type =
+                        if pk_prefix_value.size() == 0 && is_full_range(&next_col_bounds) {
+                            unreachable!()
+                        } else if pk_prefix_value.size() == pk_len {
+                            let row = {
+                                keyspace.state_store().wait_epoch(source.epoch).await?;
+                                table.get_row(&pk_prefix_value, source.epoch).await?
+                            };
+                            ScanType::PointGet(row)
+                        } else {
+                            assert!(pk_prefix_value.size() < pk_len);
+                            let iter = table
+                                .batch_iter_with_pk_bounds(
+                                    source.epoch,
+                                    &pk_prefix_value,
+                                    next_col_bounds,
+                                )
+                                .await?;
+                            ScanType::RangeScan(iter)
+                        };
```

Comment:
Not sure whether this looks good. Maybe we should `wait_epoch` only once outside the loop (and use stream iter)?

Reply:
I believe it's lightweight enough and looks good to me.